File: brain/midonet-brain/src/main/java/org/midonet/brain/southbound/vtep/model/VtepModelTranslator.java
/*
 * Copyright 2014 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.midonet.brain.southbound.vtep.model;

import java.math.BigInteger;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import org.opendaylight.ovsdb.lib.notation.OvsDBMap;
import org.opendaylight.ovsdb.lib.notation.UUID;
import org.opendaylight.ovsdb.lib.table.vtep.Logical_Switch;
import org.opendaylight.ovsdb.lib.table.vtep.Mcast_Macs_Local;
import org.opendaylight.ovsdb.lib.table.vtep.Mcast_Macs_Remote;
import org.opendaylight.ovsdb.lib.table.vtep.Physical_Port;
import org.opendaylight.ovsdb.lib.table.vtep.Physical_Switch;
import org.opendaylight.ovsdb.lib.table.vtep.Ucast_Macs_Local;
import org.opendaylight.ovsdb.lib.table.vtep.Ucast_Macs_Remote;

import org.midonet.cluster.data.vtep.model.LogicalSwitch;
import org.midonet.cluster.data.vtep.model.McastMac;
import org.midonet.cluster.data.vtep.model.PhysicalPort;
import org.midonet.cluster.data.vtep.model.PhysicalSwitch;
import org.midonet.cluster.data.vtep.model.UcastMac;
import org.midonet.packets.IPv4Addr;

/**
 * Collection of translators between VTEP entities between the OVSDB schema and
 * the Midonet schema.
 */
public class VtepModelTranslator {

    public static java.util.UUID toMido(UUID uuid) {
        return java.util.UUID.fromString(uuid.toString());
    }

    // TODO: ovsdb-specific UUIDs should be eventually erradicated
    public static UUID fromMido(java.util.UUID uuid) {
        return new UUID(uuid.toString());
    }

    public static LogicalSwitch toMido(Logical_Switch ovsdbLs,
                                       java.util.UUID uuid) {
        return new LogicalSwitch(
            uuid,
            ovsdbLs.getName(),
            // NOTE: this complicated thing is necessary because for some
            // reason the tunnel_key OvsdbSet<Integer> contains an array of
            // BigInteger which breaks when trying to use an iterator.
            ovsdbLs.getTunnel_key().isEmpty()
                ? null
                :  ((BigInteger)ovsdbLs.getTunnel_key()
                                       .toArray()[0]).intValue(),
            ovsdbLs.getDescription()
        );
    }

    public static PhysicalSwitch toMido(Physical_Switch ovsdbPs,
                                        java.util.UUID uuid) {

        Set<String> ports = new HashSet<>();
        for (UUID port : ovsdbPs.getPorts()) {
            ports.add(port.toString());
        }
        Set<IPv4Addr> mgmtIps = new HashSet<>();
        for (String ip: ovsdbPs.getManagement_ips().delegate()) {
            mgmtIps.add(IPv4Addr.fromString(ip));
        }
        Set<IPv4Addr> tunnelIps = new HashSet<>();
        for (String ip: ovsdbPs.getTunnel_ips().delegate()) {
            tunnelIps.add(IPv4Addr.fromString(ip));
        }

        return new PhysicalSwitch(
            uuid,
            ovsdbPs.getName(),
            ovsdbPs.getDescription(),
            ports,
            mgmtIps,
            tunnelIps
        );
    }

    public static PhysicalPort toMido(Physical_Port ovsdbPort) {
        return new PhysicalPort(null,
            ovsdbPort.getName(),
            ovsdbPort.getDescription(),
            toMido(ovsdbPort.getVlan_bindings()),
            toMido(ovsdbPort.getVlan_stats()),
            ovsdbPort.getPort_fault_status()
        );
    }

    public static UcastMac toMido(Ucast_Macs_Remote ovsdbMac) {
        return UcastMac.apply(null,
            ovsdbMac.getLogical_switch().isEmpty()
            ? null : toMido(ovsdbMac.getLogical_switch().iterator().next()),
            ovsdbMac.getMac(),
            ovsdbMac.getIpaddr(),
            ovsdbMac.getLocator().isEmpty()
            ? null : toMido(ovsdbMac.getLocator().iterator().next())
        );
    }

    public static UcastMac toMido(Ucast_Macs_Local ovsdbMac) {
        return UcastMac.apply(null,
            ovsdbMac.getLogical_switch().isEmpty()
                ? null: toMido(ovsdbMac.getLogical_switch().iterator().next()),
            ovsdbMac.getMac(),
            ovsdbMac.getIpaddr(),
            ovsdbMac.getLocator().isEmpty()
                ? null : toMido(ovsdbMac.getLocator().iterator().next())
        );
    }

    public static McastMac toMido(Mcast_Macs_Local ovsdbMac) {
        return McastMac.apply(null,
            ovsdbMac.getLogical_switch().isEmpty()
                ? null : toMido(ovsdbMac.getLogical_switch().iterator().next()),
            ovsdbMac.getMac(),
            ovsdbMac.getIpaddr(),
            ovsdbMac.getLocator_set().isEmpty()
                ? null : toMido(ovsdbMac.getLocator_set().iterator().next())
        );
    }

    public static McastMac toMido(Mcast_Macs_Remote ovsdbMac) {
        return McastMac.apply(null,
            ovsdbMac.getLogical_switch().isEmpty()
                ? null : toMido(ovsdbMac.getLogical_switch().iterator().next()),
            ovsdbMac.getMac(),
            ovsdbMac.getIpaddr(),
            ovsdbMac.getLocator_set().isEmpty()
                ? null : toMido(ovsdbMac.getLocator_set().iterator().next())
        );
    }

    private static Map<Integer, java.util.UUID>
        toMido(OvsDBMap<BigInteger, UUID> map) {
        Map<Integer, java.util.UUID> intMap = new HashMap<>();
        for(Map.Entry<BigInteger, UUID> e: map.entrySet()) {
            intMap.put(e.getKey().intValue(), toMido(e.getValue()));
        }
        return intMap;
    }
}

File: brain/midonet-brain/src/main/scala/org/midonet/brain/southbound/vtep/model/VtepModelTranslator.scala
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.midonet.brain.southbound.vtep.model

import java.math.BigInteger
import java.util.UUID

import scala.collection.JavaConversions._

import org.opendaylight.ovsdb.lib.notation.OvsDBMap
import org.opendaylight.ovsdb.lib.notation.{UUID => OvsdbUUID}
import org.opendaylight.ovsdb.lib.table.vtep.Logical_Switch
import org.opendaylight.ovsdb.lib.table.vtep.Mcast_Macs_Local
import org.opendaylight.ovsdb.lib.table.vtep.Mcast_Macs_Remote
import org.opendaylight.ovsdb.lib.table.vtep.Physical_Port
import org.opendaylight.ovsdb.lib.table.vtep.Physical_Switch
import org.opendaylight.ovsdb.lib.table.vtep.Ucast_Macs_Local
import org.opendaylight.ovsdb.lib.table.vtep.Ucast_Macs_Remote

import org.midonet.cluster.data.vtep.model.LogicalSwitch
import org.midonet.cluster.data.vtep.model.McastMac
import org.midonet.cluster.data.vtep.model.PhysicalPort
import org.midonet.cluster.data.vtep.model.PhysicalSwitch
import org.midonet.cluster.data.vtep.model.UcastMac
import org.midonet.packets.IPv4Addr

/**
 * Collection of translators between VTEP entities between the OVSDB schema and
 * the Midonet schema.
 */
object VtepModelTranslator {

    // TODO: ovsdb-specific UUIDs should be eventually erradicated

    def toMido(uuid: OvsdbUUID): UUID = UUID.fromString(uuid.toString)
    def fromMido(uuid: UUID): OvsdbUUID = new OvsdbUUID(uuid.toString)

    def toMido(ovsdbLs: Logical_Switch, uuid: UUID): LogicalSwitch =
        new LogicalSwitch(
            uuid,
            ovsdbLs.getName,
            // NOTE: this complicated thing is necessary because for some
            // reason the tunnel_key OvsdbSet<Integer> contains an array of
            // BigInteger which breaks when trying to use an iterator.
            if (ovsdbLs.getTunnel_key.isEmpty) null else
                ovsdbLs.getTunnel_key.toArray()(0)
                    .asInstanceOf[BigInteger].intValue(),
            ovsdbLs.getDescription
        )

    def toMido(ovsdbPs: Physical_Switch, uuid: UUID): PhysicalSwitch = {
        val ports = ovsdbPs.getPorts.delegate().toSet[OvsdbUUID].map(toMido)
        val mgmntIps = ovsdbPs.getManagement_ips.delegate().toSet
            .map(IPv4Addr.fromString)
        val tunnelIPs = ovsdbPs.getTunnel_ips.delegate().toSet
            .map(IPv4Addr.fromString)
        PhysicalSwitch(uuid, ovsdbPs.getName, ovsdbPs.getDescription,
                       ports, mgmntIps, tunnelIPs)
    }

    def toMido(ovsdbPort: Physical_Port): PhysicalPort =
        PhysicalPort(null, ovsdbPort.getName, ovsdbPort.getDescription,
                     toMido(ovsdbPort.getVlan_bindings),
                     toMido(ovsdbPort.getVlan_stats),
                     if (ovsdbPort.getPort_fault_status == null) null
                     else ovsdbPort.getPort_fault_status.delegate().toSet)

    def toMido(ovsdbMac: Ucast_Macs_Remote): UcastMac =
        UcastMac(
            if (ovsdbMac.getLogical_switch.isEmpty) null
            else toMido(ovsdbMac.getLogical_switch.iterator().next()),
            ovsdbMac.getMac,
            ovsdbMac.getIpaddr,
            if (ovsdbMac.getLocator.isEmpty) null
                else toMido(ovsdbMac.getLocator.iterator().next())
        )

    def toMido(ovsdbMac: Ucast_Macs_Local): UcastMac =
        UcastMac(
            if (ovsdbMac.getLogical_switch.isEmpty) null
            else toMido(ovsdbMac.getLogical_switch.iterator().next()),
            ovsdbMac.getMac,
            ovsdbMac.getIpaddr,
            if (ovsdbMac.getLocator.isEmpty) null
            else toMido(ovsdbMac.getLocator.iterator().next())
        )

    def toMido(ovsdbMac: Mcast_Macs_Remote): McastMac =
        McastMac(
            if (ovsdbMac.getLogical_switch.isEmpty) null
            else toMido(ovsdbMac.getLogical_switch.iterator().next()),
            ovsdbMac.getMac,
            ovsdbMac.getIpaddr,
            if (ovsdbMac.getLocator_set.isEmpty) null
            else toMido(ovsdbMac.getLocator_set.iterator().next())
        )

    def toMido(ovsdbMac: Mcast_Macs_Local): McastMac =
        McastMac(
            if (ovsdbMac.getLogical_switch.isEmpty) null
            else toMido(ovsdbMac.getLogical_switch.iterator().next()),
            ovsdbMac.getMac,
            ovsdbMac.getIpaddr,
            if (ovsdbMac.getLocator_set.isEmpty) null
            else toMido(ovsdbMac.getLocator_set.iterator().next())
        )

    def toMido(map: OvsDBMap[BigInteger, OvsdbUUID]): Map[Integer, UUID] =
        mapAsScalaMap(map)
            .map(e => new Integer(e._1.intValue()) -> toMido(e._2)).toMap
}

